ogr/ogrsf_frmts/xodr/ogrxodrlayer.cpp

```diff
@@ -0,0 +1,119 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "ogr_api.h"
+#include "ogr_geometry.h"
+#include "ogr_xodr.h"
+
+#include "cpl_error.h"
+#include <cmath>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <typeinfo>
+
+OGRXODRLayer::OGRXODRLayer(RoadElements xodrRoadElements, std::string proj4Defn)
+    : OGRXODRLayer(xodrRoadElements, proj4Defn, false)
+{
+}
+
+OGRXODRLayer::OGRXODRLayer(RoadElements xodrRoadElements, std::string proj4Defn,
+                           bool dissolveTriangulatedSurface)
+    : roadElements(xodrRoadElements), dissolveTIN(dissolveTriangulatedSurface)
+{
+    spatialRef.importFromProj4(proj4Defn.c_str());
+    ResetReading();
+}
+
+OGRXODRLayer::~OGRXODRLayer()
+{
+    if (featureDefn)
+    {
+        featureDefn->Release();
+    }
+}
+
+void OGRXODRLayer::ResetReading()
+{
+    nNextFID = 0;
+    resetRoadElementIterators();
+}
+
+void OGRXODRLayer::resetRoadElementIterators()
+{
+    roadIter = roadElements.roads.begin();
+    referenceLineIter = roadElements.referenceLines.begin();
+
+    laneIter = roadElements.lanes.begin();
+    laneSectionIter = roadElements.laneSections.begin();
+    laneRoadIDIter = roadElements.laneRoadIDs.begin();
+    laneMeshIter = roadElements.laneMeshes.begin();
+
+    laneLinesInnerIter = roadElements.laneLinesInner.begin();
+    laneLinesOuterIter = roadElements.laneLinesOuter.begin();
+
+    roadMarkIter = roadElements.roadMarks.begin();
+    roadMarkMeshIter = roadElements.roadMarkMeshes.begin();
+
+    roadObjectIter = roadElements.roadObjects.begin();
+    roadObjectMeshesIter = roadElements.roadObjectMeshes.begin();
+
+    roadSignalIter = roadElements.roadSignals.begin();
+    roadSignalMeshesIter = roadElements.roadSignalMeshes.begin();
+}
+
+OGRTriangulatedSurface OGRXODRLayer::triangulateSurface(odr::Mesh3D mesh)
```

Comment:
you could likely save memory copies with:
```suggestion
std::unique_ptr<OGRTriangulatedSurface> OGRXODRLayer::triangulateSurface(odr::Mesh3D mesh)
```